### PR TITLE
Reverse proxy config support added

### PR DESCRIPTION
**Pull request by ashendes**
**Description:**

This fix was requested by the APIM team when certain authentication flows broke upon using the gateway as a reverse proxy to IS-KM after a session hijacking fix.

This PR fixes the aforementioned issue by introducing a new reverse proxy config to replace default contexts with proxy contexts during the authorization code flow.

The introduced configuration model can be used to define a proxy context instead of the default context. Since the reported problem was with the OAuth2 consent page posting to the default /oauth2/authorize context of KM instead of proxied /authorize endpoint of the gateway, this PR addresses only this issue as an immediate fix, causing the authentication endpoint to read this configuration and use the correct endpoint to POST the consent response to [1].
Hence, [1] needs to be merged after merging this PR.

**Configuration Model:**
This configuration will not be shipped by default. Any customer using the Gateway to proxy endpoints will have to add this configuration. The use of these configurations by any pure IS customers is NOT recommended as this is not a finalized approach to the issue.

To add a reverse proxy config, add the following entry to the `deployment.toml` per config.
```
[[reverse_proxy_config]]
default_context = "<DEFAULT_PATH>"
proxy_context = "<PROXY_PATH>"
```
For example, the following configs in the toml,
```
[[reverse_proxy_config]]
default_context = "/oauth2/authorize"
proxy_context = "/authorize"

[[reverse_proxy_config]]
default_context = "/oidc/logout"
proxy_context = "/logout"
```
would lead to the addition of the following to the `identity.xml` file.
```
	<ReverseProxyConfig>
	    <ReverseProxy>
		<DefaultContext>/oauth2/authorize</DefaultContext>
		<ProxyContext>/authorize</ProxyContext> 
	    </ReverseProxy>
	    <ReverseProxy>
		<DefaultContext>/oidc/logout</DefaultContext>
		<ProxyContext>/logout</ProxyContext>
	    </ReverseProxy>
	</ReverseProxyConfig> 
```
**Expected Behavior:**
- When the configuration is in place, the configured proxy context will be used instead of the default context.
- When the configuration is not in place, the default context will be used.

[1] https://github.com/wso2/identity-apps/pull/1124

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ReverseProxyConfig.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    public String getProxyContext() {
+
+        return proxyContext;
+    }
+
+    public String getDefaultContext() {
+
+        return defaultContext;
+    }
+}
```
Comment (ruwanta) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ReverseProxyConfig.java:42]: Need new line at the end of the file

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ReverseProxyConfig.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Reverse Proxy Config.
+ */
+public class ReverseProxyConfig {
+
+    private final String defaultContext;
+    private final String proxyContext;
+
+    public ReverseProxyConfig(String defaultContext, String proxyContext) {
```
Comment (sachiniWettasinghe) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ReverseProxyConfig.java:28]: Add new line after method signature.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java [hunk 3]
```diff
@@ -440,6 +448,51 @@ public void buildLegacyFeatureConfig() {
...
+                    } else {
+                        log.warn("Configured <ReverseProxy> element at index: " + reverseProxyConfigElementIndex +
+                                " contains invalid entry.");
+                    }
+                }
+            }
+        }
+    }
+
+
```
Comment (sachiniWettasinghe) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java:495]: Remove redundant line

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 4]
```diff
@@ -237,6 +238,26 @@ public void testReadEventListenerProperty() throws Exception {
...
+                {"/oauth2/authorize", "", "/oidc/logout", "/logout", "/oauth2/authorize", "/oauth2/authorize"},
+                {"", "/authorize", "/oidc/logout", "/logout", "/oauth2/authorize", "/oauth2/authorize"},
+                {"/oauth2/authorize", "/authorize", "/oidc/logout", "/logout", "", ""},
+                {"/oauth2/authorize", "/authorize", "/oidc/logout", "/logout", null, null},
+                {"/oauth2/authorize", "/authorize", "/oidc/logout", "/logout", "/oidc/logout", "/logout"},
+                {null, null, "/oidc/logout", "/logout", "/oidc/logout", "/logout"}
+        };
+    }
+
+
```
Comment (sachiniWettasinghe) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java:260]: Remove redundant line

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java [hunk 5]
```diff
@@ -55,6 +56,7 @@
     private static Map<IdentityEventListenerConfigKey, IdentityEventListenerConfig> eventListenerConfiguration = new HashMap();
     private static Map<IdentityCacheConfigKey, IdentityCacheConfig> identityCacheConfigurationHolder = new HashMap();
     private static Map<String, IdentityCookieConfig> identityCookieConfigurationHolder = new HashMap<>();
+    private static Map<String, ReverseProxyConfig> reverseProxyConfigurationHolder = new HashMap();
```
Comment (pulasthi7) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityConfigParser.java:59]:
```suggestion
    private static Map<String, ReverseProxyConfig> reverseProxyConfigurationHolder = new HashMap<>();
```

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**